src/main/java/net/fabricmc/loader/launch/common/FabricLauncherBase.java

```diff
@@ -82,27 +82,29 @@ protected static void deobfuscate(String gameId, Path gameDir, Path jarFile, Fab
 					throw new RuntimeException("Could not locate Minecraft: " + jarFile + " not found");
 				}
 
-				// TODO: migrate to Path
-				File deobfJarDir = new File(gameDir.toFile(), ".fabric" + File.separator + "remappedJars" + (gameId.isEmpty() ? "" : File.separator + gameId));
-				if (!deobfJarDir.exists()) {
-					deobfJarDir.mkdirs();
+				Path deobfJarDir = gameDir.resolve(".fabric").resolve("remappedJars");
+
+				if (!gameId.isEmpty()) {
+					String versionedId = gameVersion.isEmpty() ? gameId : String.format("%s-%s", gameId, gameVersion);
+					deobfJarDir = deobfJarDir.resolve(versionedId);
+				}
+
+				if (!Files.exists(deobfJarDir)) {
+					Files.createDirectories(deobfJarDir);
 				}
 
 				// TODO: allow versioning mappings?
 				String deobfJarFilename = mappingConfiguration.getTargetNamespace() + "-" + jarFile.getFileName();
-				File deobfJarFile = new File(deobfJarDir, deobfJarFilename);
-				File deobfJarFileTmp = new File(deobfJarDir, deobfJarFilename + ".tmp");
-
-				Path deobfJarPath = deobfJarFile.toPath();
-				Path deobfJarPathTmp = deobfJarFileTmp.toPath();
+				Path deobfJarFile = deobfJarDir.resolve(deobfJarFilename);
+				Path deobfJarFileTmp = deobfJarDir.resolve(deobfJarFilename + ".tmp");
 
-				if (Files.exists(deobfJarPathTmp)) {
+				if (Files.exists(deobfJarFileTmp)) {
 					LOGGER.warn("Incomplete remapped file found! This means that the remapping process failed on the previous launch. If this persists, make sure to let us at Fabric know!");
-					Files.deleteIfExists(deobfJarPathTmp);
-					deobfJarFileTmp.delete();
+					Files.deleteIfExists(deobfJarFile);
+					Files.deleteIfExists(deobfJarFileTmp);
```

Comment:
No, this uses a tmp jar to remap, then renames it to the desired jar. If such a tmp jar exists at startup, it'll start over since remapping is assumed to have failed.

Deleting stale data sounds like a good idea though, I don't think we really need to keep more than one copy. A "last used" threshold or LRU rotation for cleanup are other options. I'll create an issue for this feature.